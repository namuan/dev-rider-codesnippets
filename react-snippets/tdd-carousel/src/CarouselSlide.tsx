import {ComponentPropsWithRef, ReactNode} from "react";
import styled from "styled-components";

const DEFAULT_IMAGE_HEIGHT = 500;

export type CarouselStyleProp = {
    imgUrl?: string,
    imgHeight?: string | number,
    description?: ReactNode,
    attribution?: ReactNode
} & ComponentPropsWithRef<"figure">;

type ImageComponentProps = {
    $height?: CarouselStyleProp["imgHeight"]
}

const ScaledImg = styled.img<ImageComponentProps>`
    object-fit: cover;
    width: 100%;
    height: ${(props) => {
        return typeof props.$height === "number" ? `${props.$height}px` : props.$height;
    }};
`;

const CarouselSlide = (
    {
        imgUrl,
        imgHeight = DEFAULT_IMAGE_HEIGHT,
        description,
        attribution,
        ...rest
    }: CarouselStyleProp
) => (
    <figure {...rest}>
        <ScaledImg src={imgUrl} alt={"something"} height={imgHeight}/>
        <figcaption data-testid="caption">
            <strong>{description}</strong> {attribution}
        </figcaption>
    </figure>
);

export default CarouselSlide;
