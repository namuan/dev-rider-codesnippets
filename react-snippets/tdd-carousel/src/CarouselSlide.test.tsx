import {render, screen} from "@testing-library/react";
import {describe, it, expect} from "vitest";
import CarouselSlide from "./CarouselSlide.tsx";

describe('CarouselSlide', () => {
    it('should render', () => {
        render(<CarouselSlide />)
        expect(screen.getByRole("figure")).toBeInTheDocument();
    });

    it('should render an img and figcaption', () => {
        render(<CarouselSlide />)
        const figure = screen.getByRole("figure");
        const img = screen.getByRole("img");
        const figcaption = screen.getByTestId("caption");

        expect(figure).toContainElement(img);
        expect(figure).toContainElement(figcaption);
    });

    it('should pass imgUrl to img', () => {
        const imgUrl = "https://example.com/image.png";
        render(<CarouselSlide imgUrl={imgUrl} />)
        const img = screen.getByRole("img");
        expect(img).toHaveAttribute("src", imgUrl);
    });

    it('should use description and attribution for figure caption', () => {
        const description = "Some image";
        const attribution = "John Doe";
        const props = {description, attribution};

        render(<CarouselSlide {...props} />);
        const figCaption = screen.getByTestId("caption");
        expect(figCaption).toHaveTextContent(
            `${props.description} ${props.attribution}`
        );
    });

    it('should pass other properties to figure', () => {
        const props = {
            className: "my-slides",
            "data-test-name": "my-slides"
        }
        render(<CarouselSlide {...props} />)
        const figure = screen.getByRole("figure");
        expect(figure).toHaveClass(props.className);
        expect(figure).toHaveAttribute("data-test-name", props["data-test-name"])
    });

    it('should match snapshot', () => {
        render(<CarouselSlide />)
        expect(screen.getByRole("figure")).toMatchSnapshot()
    });
});
