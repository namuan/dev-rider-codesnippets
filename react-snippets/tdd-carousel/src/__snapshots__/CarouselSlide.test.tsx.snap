// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CarouselSlide > should match snapshot 1`] = `
<figure>
  <img
    alt="something"
    class="sc-FEMpB irewsS"
    height="500"
  />
  <figcaption
    data-testid="caption"
  >
    <strong />
     
  </figcaption>
</figure>
`;
